Refresh.Analyzers/ActivityGenerator.cs

```diff
@@ -0,0 +1,170 @@
+        string code = string.Empty;
+        foreach (string name in names)
+        {
+            string type = GetTypeFromName(name);
+            string typeParam = type.ToLower();
+            
+            (string idField, string _) = GetIdFieldFromName(type);
+            
+            string method = $@"
+    /// <summary>
+    /// Creates a new {name} event from a Game{type}, and adds it to the event list.
+    /// </summary>
+    public Event Create{name}Event(GameUser userFrom, Game{type} {typeParam})
+    {{
+        Event @event = new();
+        @event.EventType = EventType.{name};
+        @event.StoredDataType = EventDataType.{type};
+        @event.Timestamp = GetTimestampMilliseconds();
+        @event.User = userFrom;
+
+        @event.{idField} = {typeParam}.{type}Id;
+
+        this._realm.Write(() => this._realm.Add(@event));
+        return @event;
+    }}
+";
+
+            code += method;
+        }
+
+
+        string sourceCode = $@"// <auto-generated/>
+using Refresh.GameServer.Types.Activity;
+using Refresh.GameServer.Types.UserData;
+using Refresh.GameServer.Types.Levels;
+
+namespace Refresh.GameServer.Database;
+
+public partial class RealmDatabaseContext
+{{
+{code}
+}}";
+        
+        context.AddSource("RealmDatabaseContext.Activity.CreateEvents.g.cs", SourceText.From(sourceCode, Encoding.UTF8));
+    }
+
+    private static void GenerateDataGetters(GeneratorExecutionContext context, IEnumerable<string> names)
+    {
+        string code = string.Empty;
+        
+        foreach (string name in names)
+        {
+            string type = "Game" + name;
+            
+            (string idField, string idFieldValue) = GetIdFieldFromName(name);
+            
+            string method = $@"
+    public {type}? Get{name}FromEvent(Event @event)
+    {{
+        if (@event.StoredDataType != EventDataType.{name})
+            throw new InvalidOperationException(""Event does not store the correct data type (expected {name})"");
+
+        if (@event.{idField} == null)
+            throw new InvalidOperationException(""Event was not created correctly, expected {idField} to not be null"");
+
+        return this._realm.All<{type}>()
+            .FirstOrDefault(l => l.{name}Id == @event.{idFieldValue});
+    }}
+";
+
+            code += method;
+        }
+        
+        string sourceCode = $@"// <auto-generated/>
+using Refresh.GameServer.Types.Activity;
+using Refresh.GameServer.Types.UserData;
+using Refresh.GameServer.Types.Levels;
+
+namespace Refresh.GameServer.Database;
+
+#nullable enable
+
+public partial class RealmDatabaseContext
+{{
+{code}
+}}";
+        
+        context.AddSource("RealmDatabaseContext.Activity.DataGetters.g.cs", SourceText.From(sourceCode, Encoding.UTF8));
+    }
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+        // while (!Debugger.IsAttached)
+        // {
+        //     Thread.Sleep(25);
+        // }
+        //
+        // Debugger.Break();
```

Comment:
odd test code probably shouldnt still be here (also pro tip: save the output to a file [Silk.NET does this and its very helpful], dont attach a debugger during this step to see the generated output)

Reply:
Yeah, I wasn't looking for output (I can view it in rider), rather I was looking for actual information about logic. Will clean up in next commit